File: app/src/main/java/ru/alemak/studentapp/util/TimeFormat.kt
package ru.alemak.studentapp.util

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

object TimeFormat {

    /**
     * Human-readable "Обновлено …" for cache / last fetch timestamps.
     * @return null if [millis] is 0 or invalid
     */
    fun updatedAtLabel(millis: Long, nowMillis: Long = System.currentTimeMillis()): String? {
        if (millis <= 0L) return null
        val diff = (nowMillis - millis).coerceAtLeast(0L)
        val minutes = TimeUnit.MILLISECONDS.toMinutes(diff)
        val hours = TimeUnit.MILLISECONDS.toHours(diff)

        val relative = when {
            minutes < 1L -> "только что"
            minutes < 60L -> "$minutes мин назад"
            hours < 24L -> {
                val h = hours.toInt()
                when {
                    h == 1 -> "1 час назад"
                    h in 2..4 -> "$h часа назад"
                    else -> "$h часов назад"
                }
            }
            else -> {
                val time = SimpleDateFormat("HH:mm", Locale("ru")).format(Date(millis))
                val day = when {
                    isSameDay(millis, nowMillis) -> "сегодня"
                    isYesterday(millis, nowMillis) -> "вчера"
                    else -> SimpleDateFormat("d MMM", Locale("ru")).format(Date(millis))
                }
                return "Обновлено: $day, $time"
            }
        }
        return "Обновлено: $relative"
    }

    private fun isSameDay(a: Long, b: Long): Boolean {
        val ca = Calendar.getInstance().apply { timeInMillis = a }
        val cb = Calendar.getInstance().apply { timeInMillis = b }
        return ca.get(Calendar.YEAR) == cb.get(Calendar.YEAR) &&
            ca.get(Calendar.DAY_OF_YEAR) == cb.get(Calendar.DAY_OF_YEAR)
    }

    private fun isYesterday(millis: Long, nowMillis: Long): Boolean {
        val y = Calendar.getInstance().apply {
            timeInMillis = nowMillis
            add(Calendar.DAY_OF_YEAR, -1)
        }
        val c = Calendar.getInstance().apply { timeInMillis = millis }
        return y.get(Calendar.YEAR) == c.get(Calendar.YEAR) &&
            y.get(Calendar.DAY_OF_YEAR) == c.get(Calendar.DAY_OF_YEAR)
    }
}

File: ios/StudentKGTU_IOS/Services/TimeFormat.swift
import Foundation

enum TimeFormat {
    static func updatedAtLabel(millis: Double, now: Date = Date()) -> String? {
        guard millis > 0 else { return nil }
        let date = Date(timeIntervalSince1970: millis / 1000)
        let diff = max(0, now.timeIntervalSince(date))
        let minutes = Int(diff / 60)
        let hours = Int(diff / 3600)

        let relative: String
        if minutes < 1 {
            relative = "только что"
        } else if minutes < 60 {
            relative = "\(minutes) мин назад"
        } else if hours < 24 {
            switch hours {
            case 1: relative = "1 час назад"
            case 2...4: relative = "\(hours) часа назад"
            default: relative = "\(hours) часов назад"
            }
        } else {
            let df = DateFormatter()
            df.locale = Locale(identifier: "ru_RU")
            df.dateFormat = "HH:mm"
            let time = df.string(from: date)
            let cal = Calendar.current
            if cal.isDateInToday(date) {
                return "Обновлено: сегодня, \(time)"
            } else if cal.isDateInYesterday(date) {
                return "Обновлено: вчера, \(time)"
            } else {
                df.dateFormat = "d MMM"
                return "Обновлено: \(df.string(from: date)), \(time)"
            }
        }
        return "Обновлено: \(relative)"
    }
}
